### PR TITLE
Initial vhost implementation

### PR DESCRIPTION
**Pull request by arronwy**
This is a sync of the Cloud Hypervisor dragonball branch for the vhost crate.
It is a superset of PR #2 with additional fixes and enhancements coming from the Cloud Hypervisor project. This PR is an exact mirror of the current vhost crate used by the Cloud Hypervisor vhost implementation.

Fixes: #8

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -3,6 +3,6 @@ name = "vhost"
 version = "0.1.0"
 authors = ["Liu Jiang <gerry@linux.alibaba.com>"]
 repository = "https://github.com/rust-vmm/vhost"
-license = "Apache-2.0"
+license = "Apache-2.0 or BSD-3-Clause"
```
Comment (andreeaflorescu) [inline Cargo.toml]: Commit 97075d1cb2fe9efe8dab91720e750b3ab9c3cf9e is not signed.

File: README.md [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
-A crate to implement the vHost ioctl interfaces and vHost-user protocol.
+A crate to support vhost backend drivers for virtio devices.
 
-## vHost Ioctl Interfaces
+## Kernel-based vHost Backend Drivers
 The vhost drivers in Linux provide in-kernel virtio device emulation. Normally the hypervisor userspace process emulates I/O accesses from the guest. Vhost puts virtio emulation code into the kernel, taking hypervisor userspace out of the picture. This allows device emulation code to directly call into kernel subsystems instead of performing system calls from userspace. The hypervisor relies on ioctl based interfaces to control those in-kernel vhost drivers, such as vhost-net, vhost-scsi and vhost-vsock etc.
 
-## vHost-user Protocol
-The vHost-user protocol is aiming to complement the ioctl interface used to control the vhost implementation in the Linux kernel. It implements the control plane needed to establish virtqueue sharing with a user space process on the same host. It uses communication over a Unix domain socket to share file descriptors in the ancillary data of the message.
+## vHost-user Backend Drivers
```
Comment (andreeaflorescu) [inline README.md]: The readme should have only 80 chars per line.

File: Cargo.toml [hunk 3]
```diff
@@ -8,7 +8,7 @@ license = "Apache-2.0 or BSD-3-Clause"
 [features]
 default = []
 vhost-vsock = []
-vhost-kern = ["vm-memory"]
+vhost-kern = ["vm-memory/backend-mmap"]
```
[the remark this answers is not in the document]
Comment (arronwy) [reply Cargo.toml]: Yes, it is not required, fixed now

File: src/vhost_kern/mod.rs [hunk 4]
```diff
@@ -1,14 +1,317 @@
...
+pub trait VhostKernBackend<'a>: AsRawFd {
+    /// Assoicated type to access guest memory.
+    type M: GuestMemory;
+
+    /// Get the object to access the guest's memory.
+    fn mem(&self) -> &Self::M;
+
+    /// Check whether the ring configuration is valid.
+    #[allow(clippy::if_same_then_else)]
+    #[allow(clippy::needless_bool)]
```
[the remark this answers is not in the document]
Comment (arronwy) [reply src/vhost_kern/mod.rs]: Yes, fixed now

File: src/vhost_kern/mod.rs [hunk 5]
```diff
@@ -1,14 +1,317 @@
...
+#[inline]
+fn ioctl_result<T>(rc: i32, res: T) -> Result<T> {
+    if rc < 0 {
+        Err(Error::IoctlError(std::io::Error::last_os_error()))
+    } else {
+        Ok(res)
+    }
+}
+
+fn guest_addr(addr: u64) -> GuestAddress {
```
[the remark this answers is not in the document]
Comment (arronwy) [reply src/vhost_kern/mod.rs]: Yes, we can use GuestAddress directly, replaced now

File: src/vhost_kern/mod.rs [hunk 6]
```diff
@@ -1,14 +1,317 @@
...
+        };
+
+        // This ioctl is called on a valid vhost fd and has its return value checked.
+        let ret = unsafe { ioctl_with_ref(self, VHOST_SET_VRING_NUM(), &vring_state) };
+        ioctl_result(ret, ())
+    }
+
+    /// Set the addresses for a given vring.
+    ///
+    /// # Arguments
```
[the remark this answers is not in the document]
Comment (arronwy) [reply src/vhost_kern/mod.rs:181]: Fixed now.

File: Cargo.toml (hunk 1 above)
Comment (arronwy) [reply Cargo.toml:6]: Fixed now.

File: README.md (hunk 2 above)
Comment (arronwy) [reply README.md:13]: Fixed now.

File: src/vhost_kern/mod.rs [hunk 7]
```diff
@@ -1,14 +1,317 @@
...
+    /// Check whether the ring configuration is valid.
+    #[allow(clippy::if_same_then_else)]
+    #[allow(clippy::needless_bool)]
+    fn is_valid(
+        &self,
+        queue_max_size: u16,
+        queue_size: u16,
+        desc_addr: GuestAddress,
+        avail_addr: GuestAddress,
+        used_addr: GuestAddress,
```
[the remark this answers is not in the document]
Comment (arronwy) [reply src/vhost_kern/mod.rs]: Yes, I have replaced to use VringConfigData for is_valid check function.

File: src/vhost_kern/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (arronwy) [reply src/vhost_kern/mod.rs:181]: Yes, fixed now.

**Comment by arronwy** [comment]
@jiangliu @sameo  I have refreshed this PR, and the main source code diff compared with origin code is as below, I found avail_addr and used_addr have the wrong order in origin code, since current code is very stable, I not sure change it to the righ order is the right fix?
```
diff -uNr ../../vhost/src/vhost_kern/mod.rs src/vhost_kern/mod.rs
--- ../../vhost/src/vhost_kern/mod.rs   2020-03-12 02:00:39.892763695 +0000
+++ src/vhost_kern/mod.rs       2020-09-03 07:24:48.373191781 +0000

     /// Check whether the ring configuration is valid.
     fn is_valid(
         &self,
-        queue_max_size: u16,
-        queue_size: u16,
-        desc_addr: GuestAddress,
-        avail_addr: GuestAddress,
-        used_addr: GuestAddress,
+        config_data: &VringConfigData
     ) -> bool {
+        let queue_size = config_data.queue_size;
         let desc_table_size = 16 * u64::from(queue_size) as GuestUsize;
         let avail_ring_size = 6 + 2 * u64::from(queue_size) as GuestUsize;
         let used_ring_size = 6 + 8 * u64::from(queue_size) as GuestUsize;
-        if queue_size > queue_max_size || queue_size == 0 || (queue_size & (queue_size - 1)) != 0 {
+        if queue_size > config_data.queue_max_size || queue_size == 0 || (queue_size & (queue_size - 1)) != 0 {
             false
-        } else if desc_addr
+        } else if GuestAddress(config_data.desc_table_addr)
             .checked_add(desc_table_size)
             .map_or(true, |v| !self.mem().address_in_range(v))
         {
             false
-        } else if avail_addr
+        } else if GuestAddress(config_data.avail_ring_addr)
             .checked_add(avail_ring_size)
             .map_or(true, |v| !self.mem().address_in_range(v))
         {
             false
-        } else if used_addr
+        } else if GuestAddress(config_data.used_ring_addr)
             .checked_add(used_ring_size)
             .map_or(true, |v| !self.mem().address_in_range(v))
         {
@@ -188,42 +179,30 @@
     fn set_vring_addr(&mut self, queue_index: usize, config_data: &VringConfigData) -> Result<()> {
-        if !self.is_valid(
-            config_data.queue_max_size,
-            config_data.queue_size,
-            guest_addr(config_data.desc_table_addr),
-            guest_addr(config_data.used_ring_addr),
-            guest_addr(config_data.avail_ring_addr),
-        ) {
+        if !self.is_valid(config_data) {
             return Err(Error::InvalidQueue);
         }
```

**Comment by jiangliu** [comment]
> @jiangliu @sameo I have refreshed this PR, and the main source code diff compared with origin code is as below, I found avail_addr and used_addr have the wrong order in origin code, since current code is very stable, I not sure change it to the righ order is the right fix?

That‘s a bug, your fix is correct.

**Comment by andreeaflorescu** [comment]
@arronwy what about the CI? Are you planning on adding it in a subsequent PR? Lemme know if you need help with that. It looks like there are a few features as well which are not default and they'll not be tested by the default pipeline in the rust-vmm-ci, so that means that a custom pipeline would also be needed.

**Comment by arronwy** [comment]
> @arronwy what about the CI? Are you planning on adding it in a subsequent PR? Lemme know if you need help with that. It looks like there are a few features as well which are not default and they'll not be tested by the default pipeline in the rust-vmm-ci, so that means that a custom pipeline would also be needed.

Thanks, I'll follow the readme in rust-vmm-ci project and submit the CI patches.

**Comment by arronwy** [comment]
> @arronwy what about the CI? Are you planning on adding it in a subsequent PR? Lemme know if you need help with that. It looks like there are a few features as well which are not default and they'll not be tested by the default pipeline in the rust-vmm-ci, so that means that a custom pipeline would also be needed.

I added a commit to do the step 1,2 in rust-vmm-ci, for step 3 Create a new pipeline definition in Buildkite, we may need the admin to create it first.

**Comment by andreeaflorescu** [comment]
> > @arronwy what about the CI? Are you planning on adding it in a subsequent PR? Lemme know if you need help with that. It looks like there are a few features as well which are not default and they'll not be tested by the default pipeline in the rust-vmm-ci, so that means that a custom pipeline would also be needed.
> 
> I added a commit to do the step 1,2 in rust-vmm-ci, for step 3 Create a new pipeline definition in Buildkite, we may need the admin to create it first.

I created the Buildkite pipeline for you and also configured the webhook. I believe we need a force push though to trigger the build :(

**Comment by slp** [comment]
Thanks @arronwy. I'm happy with this initial implementation. The test coverage is a bit low, but we can improve on this later.

There are two CI tests failing. One is the `unittest` pipeline for the `musl-arm`, which can be fixed by using [this workaround](https://github.com/rust-vmm/event-manager/commit/907511a41fe8b60b1156a6aa9f23bf416c14ef67) (thanks to @andreeaflorescu for pointing me to that). The other is `commit-format`, because some commit title exceeds 50 chars, but I think we can give it a pass this time.

**Comment by arronwy** [comment]
> Thanks @arronwy. I'm happy with this initial implementation. The test coverage is a bit low, but we can improve on this later.
> 
> There are two CI tests failing. One is the `unittest` pipeline for the `musl-arm`, which can be fixed by using [this workaround](https://github.com/rust-vmm/event-manager/commit/907511a41fe8b60b1156a6aa9f23bf416c14ef67) (thanks to @andreeaflorescu for pointing me to that). The other is `commit-format`, because some commit title exceeds 50 chars, but I think we can give it a pass this time.

Thanks @slp @andreeaflorescu for your help and patch, finally all checks have passed.

**Review by slp — APPROVED** [review]

**Review by jiangliu — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]
I haven't really looked at the code after it was updated, but it looks like all my previous comments were resolved.

Thanks for upstreaming this!

**Pull request merged**